pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,320 @@
+	files := []string{}
+	for _, file := range oldFiles {
+		files = append(files, filepath.Join("/proc", fmt.Sprintf("%d", pid), "root", file))
+	}
+	return files
+}
+
+// Run the main watchdog loop.
+func (o *FileWatcherOptions) Run(ctx context.Context) error {
+	for {
+		{
+			instanceCtx, shutdown := context.WithCancel(ctx)
+			defer shutdown()
+			select {
+			case <-ctx.Done():
+				// exit(0)
+				shutdown()
+				return nil
+			default:
+			}
+			if err := o.runWatchdog(instanceCtx); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// runWatchdog run single instance of watchdog.
+func (o *FileWatcherOptions) runWatchdog(ctx context.Context) error {
+	watchdogCtx, shutdown := context.WithCancel(ctx)
+	defer shutdown()
+
+	// Handle watchdog shutdown
+	go func() {
+		defer shutdown()
+		<-ctx.Done()
+	}()
+
+	pidObservedCh := make(chan int)
+	go o.runPidObserver(watchdogCtx, pidObservedCh)
+
+	// Wait while we get the initial PID for the process
+	klog.Infof("Waiting for process %q PID ...", o.ProcessName)
+	currentPID := <-pidObservedCh
+
+	// Mutate path for specified files as '/proc/PID/root/<original path>'
+	// This means side-car container don't have to duplicate the mounts from main container.
+	// This require shared PID namespace feature.
+	filesToWatch := o.addProcPrefixToFilesFn(o.Files, currentPID)
+	klog.Infof("Watching for changes in: %s", spew.Sdump(filesToWatch))
+
+	// Read initial file content. If shared PID namespace does not work, this will error.
+	initialContent, err := readInitialFileContent(filesToWatch)
+	if err != nil {
+		// TODO: remove this once we get aggregated logging
+		o.recorder.Warningf("FileChangeWatchdogFailed", "Reading initial file content failed: %v", err)
+		return fmt.Errorf("unable to read initial file content: %v", err)
+	}
+
+	o.recorder.Eventf("FileChangeWatchdogStarted", "Started watching files for process %s[%d]", o.ProcessName, currentPID)
+
+	observer, err := fileobserver.NewObserver(o.Interval)
+	if err != nil {
+		o.recorder.Warningf("ObserverFailed", "Failed to start to file observer: %v", err)
+		return fmt.Errorf("unable to start file observer: %v", err)
+	}
+
+	observer.AddReactor(func(file string, action fileobserver.ActionType) error {
+		retries := 0
+		// When file change is observed, don't give up on errors when we fail to send signal, but retry sending
+		// and fail hard if we fail to deliver it. Try 5 times, sleep 1s between attempts.
+		var lastSignalErr error
+		pollErr := wait.PollImmediate(1*time.Second, 5*time.Second, func() (done bool, err error) {
+			// we already signalled this PID to terminate and the process is being gracefully terminated now.
+			// do not send the signal to the same PID twice, but wait for the new PID to appear.
+			if currentPID == o.lastUsedPid {
+				return true, nil
+			}
+			defer func() {
+				shutdown()
+			}()
+			o.recorder.Eventf("FileChangeObserved", "Observed change in file %q, sending SIGINT to %s[%d]", file, o.ProcessName, currentPID)
```

Comment:
INT or TERM?

Reply:
not that the API server will terminate non-gracefully on the second signal. Does this mean it has only 1 second to terminate gracefully?

Reply:
I added graceful shutdown sequence